### PR TITLE
Issue #109

### PR DESCRIPTION
**Pull request by burrbd**
Fix for https://github.com/gorilla/mux/issues/109.

Unfortunately we broke the query matcher when the query's ```defaultPattern``` was changed in https://github.com/gorilla/mux/pull/108.

In this PR the ```getUrlQuery``` method loops over query parameters and only returns a string to match if the the parameter exists in the request (which is how it should have been in the first place).

### CODE REVIEW

File: mux_test.go [hunk 1]
```diff
@@ -606,6 +624,24 @@ func TestQueries(t *testing.T) {
 			path:        "",
 			shouldMatch: false,
 		},
+		{
+			title:       "Queries route with no parameter in request, should not match",
```
Comment (burrbd) [inline mux_test.go]: This is the test that should cover #109.

File: regexp.go [hunk 2]
```diff
@@ -186,9 +186,13 @@ func (r *routeRegexp) getUrlQuery(req *http.Request) string {
 	if !r.matchQuery {
 		return ""
 	}
-	key := strings.Split(r.template, "=")[0]
-	val := req.URL.Query().Get(key)
-	return key + "=" + val
+	templateKey := strings.Split(r.template, "=")[0]
```
Comment (kisielk) [inline regexp.go]: use `SplitN` here with a max of 2

Comment (kisielk) [inline regexp.go]: I just did this quickly :)

**Comment by kisielk** [comment]
Thanks for the fix.

It really seems unnecessary to put the key and = in to the template now, I have to get around to refactoring this code and creating a dedicated query matcher which just has a key and value field to avoid all of this mess.

**Comment by burrbd** [comment]
Was a bad thing to break. Hopefully those extra tests will help with any refactoring. It would be nice if the refactoring could handle ```foo=1&foo=2&foo=3```. Might need a new queries variable along with ```mux.Vars(request)``` to do this.

**Comment by kisielk** [comment]
You could already extract that from `r.URL.Queries()["foo"]` inside your handler if that's what you need...

**Comment by burrbd** [comment]
True. But it's not great that in this case ```vars["foo"] ``` would be equal to a string (the first value) as opposed to a slice.

And the matcher should test the value of each ```&foo``` rather than just one item ... I think? O_o

**Comment by kisielk** [comment]
Yeah, it's unfortunate about vars but we can't change the API.

But yes, I the matcher should check each foo and just return the first match, which is what it's doing now.

**Comment by kisielk** [comment]
Actually, that's not quite right.. it's only attempting the match on the first query... *sigh*. It seems like we'd probably have to do a fair bit of refactoring to get the desired behaviour with multiple query items with the same name.

**Pull request merged**